modules/neopixel/neopx.py

```
@@ -294,8 +383,125 @@ def animate(self, identifiers, color, animation):
         self.animation = True
         if animation in animations:
             self.thread = threading.Thread(target=animations[animation], args=(identifiers, color,))
-        self.thread.start()
+            self.thread.start()
+
+    def handle_animate(self, animation, color=None, color2=None, repeat=1):
+        """
+        Handle animation requests from pubsub
+        :param animation: animation name
+        :param color: optional first color
+        :param color2: optional second color
+        :param repeat: optional repeat count (1-10)
+        """
+        if self.protocol == 'ESP32':
+            self.send_animation_command(animation, color, color2, repeat)
+        else:
+            # Fall back to standard animation system
+            color_val = color if color else NeoPx.COLOR_RED
+            self.animate(self.all, color_val, animation.lower())
 
+    def send_animation_command(self, animation, color=None, color2=None, repeat=1):
+        """
+        Send animation command to ESP32
+        :param animation: animation name
+        :param color: optional first color (string or RGB tuple)
+        :param color2: optional second color (string or RGB tuple)
+        :param repeat: optional repeat count (1-10)
+        """
+        if self.protocol != 'ESP32':
```

I would expect an error here, as they're trying to do something they shouldn't, rather than failing silently as this makes it very difficult to debug.

---

I didn't quite understand what you said here.

---

Don't just return if there is an error. Throw an exception.